### PR TITLE
tests: Add a test construct of a simple stencil dialect use in CI

### PR DESCRIPTION
**Pull request by georgebisbas**
This test aims to keep track of printing stencil dialect operations after forming a stencil computation.
No weights, just accesses.

### CODE REVIEW

**Review by PapyChacal — APPROVED** [review]

**Review by webmiche — COMMENTED** [review]
IMO, this is getting into the domain of integration tests, which feels like it should be separated than how it is done right now. Though this is a major refactor of the testing infrastructure, certainly not something for this PR. Just my two cents :)

File: tests/dialects/test_stencil.py [hunk 1]
```diff
@@ -560,3 +570,74 @@ def test_stencil_access():
...
+def test_1d3pt_stencil_construct():
+    """
+    An example 1d-3pt stencil implementation from XDSL Python-land
+    using the stencil dialect
+    """
+
+    shape = (8,)
+    space_order = 2
+    r = space_order // 2
+    symbolic_shape = (0 - r, shape[0] - r)
```
Comment (AntonLydike) [inline tests/dialects/test_stencil.py:585]: shouldn't this be `shape[0] + r`?

File: tests/dialects/test_stencil.py [hunk 2]
```diff
@@ -560,3 +570,74 @@ def test_stencil_access():
...
+    # Stencil computation
+    stencil_ops = [
+        stencil_acs_l := AccessOp.get(block0.args[0], (-1,)),
+        stencil_acs_c := AccessOp.get(block0.args[0], (0,)),
+        stencil_acs_r := AccessOp.get(block0.args[0], (1,)),
+        stencil_comp0 := Addf(stencil_acs_l, stencil_acs_c),
+        stencil_comp1 := Addf(stencil_comp0, stencil_acs_r),
+        # Define the return operation
+        return_op_ := ReturnOp.get([stencil_comp1]),  # type: ignore
+    ]
```
Comment (AntonLydike) [inline tests/dialects/test_stencil.py:605]: I would suggest we use the implicit builder here?

**Review by AntonLydike — COMMENTED** [review]
This looks alright so far, I think we should add a lot more comments, what exactly each line does. We should also use the implicit builder to make the code a lot cleaner.

File: tests/dialects/test_stencil.py [hunk 3]
```diff
@@ -560,3 +570,74 @@ def test_stencil_access():
...
+
+    # Stencil computation
+    stencil_ops = [
+        stencil_acs_l := AccessOp.get(block0.args[0], (-1,)),
+        stencil_acs_c := AccessOp.get(block0.args[0], (0,)),
+        stencil_acs_r := AccessOp.get(block0.args[0], (1,)),
+        stencil_comp0 := Addf(stencil_acs_l, stencil_acs_c),
+        stencil_comp1 := Addf(stencil_comp0, stencil_acs_r),
+        # Define the return operation
+        return_op_ := ReturnOp.get([stencil_comp1]),  # type: ignore
```
Comment (superlopuh) [inline tests/dialects/test_stencil.py:604]: Why the type ignore?

File: tests/dialects/test_stencil.py (hunk 1 above)
Comment (PapyChacal) [reply tests/dialects/test_stencil.py:585]: Here the shape is the halo including size! Maybe the halo excluding size could be clearer, I'm not sure;
Though this is what George came up with and he's amongst the target stencil experts 😛

File: tests/dialects/test_stencil.py (hunk 3 above)
Comment (PapyChacal) [reply tests/dialects/test_stencil.py:604]: Because otherwise pyright complains about the unused variable, but declaring one felt more consistent and readable

Comment (superlopuh) [reply tests/dialects/test_stencil.py:604]: But it should probably go away if this is migrated to the builder, which I would also recommend :)

Comment (PapyChacal) [reply tests/dialects/test_stencil.py:604]: Yeah I would use the builder for this either way too; are the new with blocks merged already?
(Otherwise I like the underscore suggestion, thanks!)

Comment (superlopuh) [reply tests/dialects/test_stencil.py:604]: The with blocks are merged, and are probably the thing to use. I'm still trying to work out ergonomics for optional regions

**Review by meshtag — APPROVED** [review]
I like this idea of having stencils in the Jupyter notebook!

File: tests/dialects/test_stencil.py (hunk 3 above)
Comment (georgebisbas) [reply tests/dialects/test_stencil.py:657]: @superlopuh could you give me a pointer on that?

File: tests/dialects/test_stencil.py (hunk 1 above)
Comment (georgebisbas) [reply tests/dialects/test_stencil.py:638]: What we need here is the correct input in the FieldType, i.e. the computational domain size. So I guess Anton is correct. Please correct me if I am wrong. @PapyChacal, you are the stencil dialect expert. What is the correct way to put this here?

File: tests/dialects/test_stencil.py (hunk 2 above)
Comment (georgebisbas) [reply tests/dialects/test_stencil.py]: To build what?

File: tests/dialects/test_stencil.py (hunk 3 above)
Comment (superlopuh) [reply tests/dialects/test_stencil.py]: The best examples are probably in the pytest pdl interpreter tests, `test_pdl_interpreter.py`

File: tests/dialects/test_stencil.py (hunk 2 above)
Comment (superlopuh) [reply tests/dialects/test_stencil.py]: the whole module, I guess, would probably be more readable

File: tests/dialects/test_stencil.py (hunk 1 above)
Comment (PapyChacal) [reply tests/dialects/test_stencil.py:638]: The fieldtype is none of that, it's just the buffer size. (E.g. in DeVito it's compute domain + 2 in all directions, even where the input with halo is +1)
What is expressed here is an input buffer bounded by [-1,7], which yields size 8, which is what I read by shape = (8,) personally.

You set the store bounds to [0,6], those will be the compute bounds here. The offsets are -1/+1 so the buffer size is exactly the minimum needed.

So, this is a correct stencil computation, I don't know if it's what you wanted to express though

File: tests/dialects/test_stencil.py (hunk 3 above)
Comment (PapyChacal) [reply tests/dialects/test_stencil.py]: Oh no, just rewrote it with the implicit stuff, not finding the with blocks :smiling_face_with_tear: Will have a look

Comment (PapyChacal) [reply tests/dialects/test_stencil.py]: Okay, it was easy actually, done

**Comment by tobiasgrosser** [comment]
This seems to be inactive. I close it for now. Please reopen it in case you feel this is still relevant and should be updated and integrated.

**Comment by georgebisbas** [comment]
This was updated last week, and rebased on top of master, also asked for reviews.

**Comment by tobiasgrosser** [comment]
Uh, sorry. @PapyChacal, is there a chance you could have a look? What is the status with this PR?

**Comment by PapyChacal** [comment]
Right, I'll have a look now

File: stencil1d.mlir
Comment (webmiche) [inline stencil1d.mlir:1]: Is this file required to be in the root directory of the repo? Looks a bit weird to have .mlir files there :thinking: 
@georgebisbas @PapyChacal

Comment (georgebisbas) [reply stencil1d.mlir:1]: Thanks for spotting this! What an oversight!

Fixing here: https://github.com/xdslproject/xdsl/pull/1871
